planner/core/push_down_predicate.go

```diff
@@ -0,0 +1,249 @@
+ * 		  For example, JSON functions.(To be supplemented)
+ */
+func withHeavyCostFunction(cond expression.Expression) bool {
+	if binop, ok := cond.(*expression.ScalarFunction); ok {
+		switch binop.FuncName.L {
+		// JSON functions
+		case ast.JSONType, ast.JSONExtract, ast.JSONUnquote, ast.JSONValid, ast.JSONSearch, ast.JSONContains, ast.JSONContainsPath, ast.JSONKeys, ast.JSONLength, ast.JSONDepth, ast.JSONQuote, ast.JSONSet, ast.JSONInsert, ast.JSONReplace, ast.JSONRemove, ast.JSONMerge, ast.JSONArrayAppend, ast.JSONArrayInsert, ast.JSONMergePatch, ast.JSONMergePreserve, ast.JSONPretty, ast.JSONStorageSize:
+			return true
+			// TODO: add more heavy cost functions
+		}
+	}
+	return false
+}
+
+/**
+ * @brief: This function is used to remove the conditions that needed to be pushed down.
+ * @param: physicalSelection: the PhysicalSelection to be modified
+ * @param: exprs: the conditions to be removed
+ */
+func removeSpecificExprsFromSelection(physicalSelection *PhysicalSelection, exprs []expression.Expression) {
+	newConds := make([]expression.Expression, 0, len(physicalSelection.Conditions))
+	for _, cond := range physicalSelection.Conditions {
+		if !expression.Contains(exprs, cond) {
+			newConds = append(newConds, cond)
+		}
+	}
+	physicalSelection.Conditions = newConds
+}
+
+/**
+ * @brief: This function is used to push down the some filter conditions of the selection to the tablescan.
+ * @param: sctx: the session context
+ * @param: physicalSelection: the PhysicalSelection containing the conditions to be pushed down
+ * @param: physicalTableScan: the PhysicalTableScan to be pushed down to
+ */
+func predicatePushDownToTableScanImpl(sctx sessionctx.Context, physicalSelection *PhysicalSelection, physicalTableScan *PhysicalTableScan) {
+	conds := physicalSelection.Conditions
+	if len(conds) == 0 {
+		return
+	}
+
+	// group the conditions by columns and sort them by selectivity
+	sortedConds := groupByColumnsSortBySelectivity(sctx, conds, physicalTableScan)
+
+	selectedConds := make([]expression.Expression, 0, len(conds))
+	selectedIncome := 0.0
+	selectedColumnCount := 0
+	totalColumnCount := len(physicalTableScan.Columns)
+
+	for _, exprGroup := range sortedConds {
+		mergedConds := append(selectedConds, exprGroup.exprs...)
+		selectivity, _, err := physicalTableScan.tblColHists.Selectivity(sctx, mergedConds, nil)
+		if err != nil {
+			logutil.BgLogger().Debug("calculate selectivity failed, do not push down the conditions group", zap.Error(err))
+			continue
+		}
+		var cols []*expression.Column
+		cols = expression.ExtractColumnsFromExpressions(cols, mergedConds, nil)
+		colCnt := len(cols)
+		income := (1 - selectivity) * (float64(totalColumnCount) - float64(colCnt))
+		// If selectedColumnCount does not change,
+		// or the income increase larger than the threshold after pushing down the group, push down it.
+		if colCnt == selectedColumnCount || income-selectedIncome > incomeImproveThreshold {
```

Comment:
There are several factors that may need to be considered when assessing benefits:
1. The complexity of a single predicate
2. Maximum number of delayed materialization predicates
2. The types of the remaining columns, such as string and json, have different reading speeds for different types.

Reply:
I will do more perf test to adjust the algorithm.